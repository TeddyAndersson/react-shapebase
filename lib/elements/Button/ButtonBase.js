import React from 'react';
import { findDOMNode } from 'react-dom';
import _ from 'lodash';
import keycode from 'keycode';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import colors from '../../styles/colors'

class Base extends React.Component{
    constructor() {
        super();
        this.state = {
            button: null,
        }
    }

    componentDidMount() {
        this.setState({
            button: findDOMNode(this)
        })
        // listenForFocusKeys();
    }

    componentWillUnmount() {
        this.setState({
            button: null
        })
    }


    handleClick = (e) => {
        const { disabled, loading } = this.props

        if (disabled || loading) {
          e.preventDefault()
          return
        }
    
        _.invoke(this.props, 'onClick', e, this.props)
    }
    render(){
        const {
            auto,
            children,
            classes,
            className,
            component,
            disabled,
            loading,
            onClick,
            rootRef,
            type,
            secondary,
            primary,
            ...other
        } = this.props;

        const buttonProps = {

        };

        let ComponentProp = component;

        if (ComponentProp === 'button') {
            buttonProps.type = type || 'button';
        }

        if (ComponentProp !== 'a') {
            buttonProps.role = buttonProps.role || 'button';
            buttonProps.disabled = disabled;
        }

        if (loading) {
            buttonProps.disabled = true;
        }

        return (
            <ComponentProp
                onClick={this.handleClick}
                className={className}
                {...buttonProps}
                {...other}
            >   
                {children}
            </ComponentProp>
        )
    };
};

Base.propTypes = {
    /**
     * The content of the component.
     */
    auto: PropTypes.bool,

    /**
     * The content of the component.
     */
    children: PropTypes.node,
    /**
     * Useful to extend the style applied to components.
     */
    /**
     * @ignore
     */
    className: PropTypes.string,
    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     * The default value is a `button`.
     */
    component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
    
    /**
     * If `true`, the base button will be disabled.
     */
    disabled: PropTypes.bool,

    /**
     * The CSS class applied while the component is keyboard focused.
     */
    keyboardFocusedClassName: PropTypes.string,

    /**
     * @ignore
     */
    onBlur: PropTypes.func,

    /**
     * @ignore
     */
    onClick: PropTypes.func,

    /**
     * @ignore
     */
    role: PropTypes.string,

    /**
     * Use that property to pass a ref callback to the root component.
     */
    rootRef: PropTypes.func,

    /**
     * @ignore
     */
    type: PropTypes.string,

    /**
     * @ignore
     */
    primary: PropTypes.bool,

    /**
     * @ignore
     */
    secondary: PropTypes.bool,

  };

  Base.defaultProps = {
    secondary: false,
    primary: false,
  }

/** Styling the Component with styled-components */
const ButtonBase = styled(Base)`
    margin: 0
    border: 0
    padding: 0
    display: inline-block
    vertical-align: middle
    white-space: normal
    background: none
    line-height: 1
    font-size: inherit
    font-family: inherit
    user-select: none;
    display:flex;
    justify-content: center;
    align-items: center;
    background: #eee;
    color: #555;
    font-weight: bold;
    cursor: pointer;

  
  ${props => props.dense && css`
    height: 1.8em;
    min-height: 1em;
  `}

  ${props => props.auto && css`
    width: auto;
    min-width: 0;
  `}

  ${props => props.primary && css`
    background: ${colors.info};
    color: #fff;
  `}

  ${props => props.secondary && css`
    background: ${colors.darkBlue};
    color: #fff;
  `}

  ${props => props.accent && css`
    background: ${colors.darkBlue};
    color: #fff;
  `}

  ${props => props.loading && css`
      color: #bdbdbd;
      cursor: not-allow;
  `}

  ${props => props.disabled && css`
    background: #f1f2f2;
    color: #bdbdbd;
    cursor: not-allow;
  `};
`;

export default ButtonBase;