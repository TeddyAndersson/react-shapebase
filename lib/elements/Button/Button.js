import React from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import ButtonBase from './ButtonBase';
import RectangleShape from '../Shape/RectangleShape';
import CircleShape from '../Shape/CircleShape';
import SquareShape from '../Shape/SquareShape';
import Loader from '../Loader/Loader';

const Button = (props) => {
  const {
    children,
    disabled,
    raised,
    shape,
    size,
    loading,
    loadingColor,
    ...other
  } = props;


  let ShapeComponent = undefined;

  if (shape === 'rectangle') {
    ShapeComponent = RectangleShape;
  }

  if (shape === 'square') {
    ShapeComponent = SquareShape;
  }

  if (shape === 'circle') {
    ShapeComponent = CircleShape;
  }

  return (
    <ButtonBase
      disabled={disabled}
      component={ShapeComponent}
      raised={raised}
      size={size}
      loading={loading}
      {...other}
    >
      <span>
        {
          loading ? 
            <Loader color={loadingColor} /> 
            : children
        }
      </span>
    </ButtonBase>
  );
};

Button.propTypes = {
  /**
    * The content of the button.
    */
  children: PropTypes.node.isRequired,
  /**
    * Useful to extend the style applied to components.
    */
  classes: PropTypes.shape({}),
  /**
    * @ignore
    */
  className: PropTypes.string,
  /**
    * The component used for the root node.
    * Either a string to use a DOM element or a component.
    * The default value is a `button`.
    */
  component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

  /**
    * If `true`, the button will be disabled.
    */
  disabled: PropTypes.bool,
  /**
    * The URL to link to when the button is clicked.
    * If defined, an `a` element will be used as the root node.
    */
   href: PropTypes.string,
   /**
    * If `true`, the button will use raised styling.
    */
  
  /**
    * Loading indicator for the button.
    */
  loading: PropTypes.bool,

  /**
    * The color of the loading spinner.
    */
  loadingColor: PropTypes.string,

  /**
    * If `true`, the button will use raised styling.
    */
  raised: PropTypes.number,
  /**
    * The size of the button.
    */
  shape: PropTypes.oneOf(['circle', 'rectangle', 'square']),
  /**
    * The size of the button.
    */
  size: PropTypes.number,
  /**
    * The color of the component.
    */
  theme: PropTypes.oneOf(['primary', 'accent', 'secondary', 'transparent']),
  /**
    * @ignore
    */
  type: PropTypes.string,
};

Button.defaultProps = {
  disabled: false,
  href: undefined,
  raised: 0,
  shape: 'rectangle',
  size: 1,
  color: 'primary',
  type: 'button',
  loading: false,
  loadingColor: "rgba(255,255,255,0.9)"
};



export default Button;