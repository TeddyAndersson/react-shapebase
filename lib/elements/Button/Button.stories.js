import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Button from './Button';
import Grid from '../Grid'

storiesOf('Button', module)
  .add('Button', () => (
    <Grid columns={1} rowGap={1}>
      <Button onClick={action('clicked')} >Default</Button>
      <Button onClick={action('clicked')} primary >Primary</Button>
      <Button onClick={action('clicked')} secondary >Secondary</Button>
      <Button loading={true} onClick={action('clicked')}>Button</Button>
      <Button onClick={action('clicked')} disabled>Disabled</Button>
    </Grid>
  ))
  .add('Raised Button', () => (
    <Grid columns={1} rowGap={1}>
      <Button raised={1} onClick={action('clicked')}>Default</Button>
      <Button raised={1} onClick={action('clicked')} primary>Primary</Button>
      <Button raised={1} onClick={action('clicked')} secondary>Secondary</Button>
      <Button raised={1} onClick={action('clicked')} disabled>Disabled</Button>
    </Grid>
  ))
  .add('Circle Button', () => (
    <Grid columns={1} rowGap={1}>
      <Button onClick={action('clicked')} shape='circle'>D</Button>
      <Button onClick={action('clicked')} shape='circle' primary>P</Button>
      <Button onClick={action('clicked')} shape='circle' secondary>S</Button>
      <Button onClick={action('clicked')} shape='circle' disabled>S</Button>
    </Grid>
  ))
  .add('Raised Circle Button', () => (
    <Grid columns={1} rowGap={1}>
      <Button raised={1} onClick={action('clicked')} shape='circle'>D</Button>
      <Button raised={1} onClick={action('clicked')} shape='circle' primary>P</Button>
      <Button raised={1} onClick={action('clicked')} shape='circle' secondary>S</Button>
      <Button raised={1} onClick={action('clicked')} shape='circle' disabled>No</Button>
    </Grid>
  ))
  .add('Square Button', () => (
    <Grid columns={1} rowGap={1}>
      <Button onClick={action('clicked')} shape='square'>D</Button>
      <Button onClick={action('clicked')} shape='square' primary>P</Button>
      <Button onClick={action('clicked')} shape='square' secondary>S</Button>
      <Button onClick={action('clicked')} shape='square' disabled>No</Button>
    </Grid>
  ))
  .add('Raised Suquare Button', () => (
    <Grid columns={1} rowGap={1}>
      <Button raised={1} onClick={action('clicked')} shape='square' >D</Button>
      <Button raised={1} onClick={action('clicked')} shape='square' primary>P</Button>
      <Button raised={1} onClick={action('clicked')} shape='square' secondary>S</Button>
      <Button raised={1} onClick={action('clicked')} shape='square' disabled>No</Button>
    </Grid>
  ));