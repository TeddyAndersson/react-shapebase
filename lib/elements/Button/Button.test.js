import React from 'react';
import Button from './Button'
import renderer from 'react-test-renderer';

test('should render Button', () => {
  const component = renderer.create(
    <Button onClick={()=>{}}>Default</Button>
);
  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
})
