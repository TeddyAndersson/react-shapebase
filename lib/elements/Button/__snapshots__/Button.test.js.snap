// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Button 1`] = `
<div
  className="sc-bdVaJa fLeRSV sc-htpNat fLfkdc sc-bwzfXH iRJfEt"
  color="primary"
  disabled={false}
  href={undefined}
  onClick={[Function]}
  role="button"
>
  <span>
    Default
  </span>
</div>
`;
